spec/ddtrace/configuration/settings_spec.rb

```diff
@@ -6,6 +6,44 @@
 RSpec.describe Datadog::Configuration::Settings do
   subject(:settings) { described_class.new }
 
+  describe '#sampling' do
+    describe '#rate_limit' do
+      subject(:rate_limit) { settings.sampling.rate_limit }
+
+      context 'default' do
+        it { is_expected.to be nil }
+      end
+
+      context 'when ENV is provided' do
+        around do |example|
+          ClimateControl.modify(Datadog::Ext::Sampling::ENV_RATE_LIMIT => '20.0') do
+            example.run
+          end
+        end
+
+        it { is_expected.to eq(20.0) }
```

Comment:
This test doesn't pass right now, because when the test suite loads, the environment variable is read and its value cached into the class definition for configuration settings. So despite changing the env var value, it ignores it and uses the value it already cached.

Although this is generally not an issue for most applications (they're expected to set ENV at load time), its problematic for our test suite. Will need to think of the best way to deal with this.

Reply:
Turns out I was wrong; the value was not cached onto the class definition. Options and their values are per-settings object, so when creating new settings objects, it should lazily load the default from the ENV after altering it in the `before` block.

Instead, there was a bug in how `settings` are instantiated (in this case for `sampling`.) Because sub-settings were just another option, and they were defined as `default settings_class.new`, it was re-using the same object for the sub-settings across all new settings objects, hence why the old value appeared to be cached. These objects should be unique, so I changed this instead to a lazy scheme to fix this.